### PR TITLE
Minor fixes to enable building with clang on OS X

### PR DESCRIPTION
**Pull request by taion**
There's 3 things going on here.

1. Minor fixes where clang is less permissive than gcc regarding syntax.
2. Some weirdness with `TupleResult` and `TupleLastResult` not properly compiling in clang, but with a workaround that I think is good enough.
3. Working around a bunch of pain with the non-standard-compliant (I think) `std::vector<bool>` specialization in Apple's libc++ that's even worse than the normal `std::vector<bool>` specialization: https://llvm.org/bugs/show_bug.cgi?id=10314

Built and tested with Xcode 6.3.1 toolchain and gcc 4.9.2.

ETA: For #55 (adding this to get the link to the PR from the issue)

### CODE REVIEW

**Comment by taion** [comment]
Revised commit only changed the commit comment to reference the issue.

File: cppwamp/include/cppwamp/internal/endian.ipp [hunk 1]
```diff
@@ -5,7 +5,7 @@
                     http://www.boost.org/LICENSE_1_0.txt)
 ------------------------------------------------------------------------------*/
 
-#include "endian.h"
+#include "endian.hpp"
```
Comment (ecorm) [inline cppwamp/include/cppwamp/internal/endian.ipp]: `#include "endian.hpp"` should actually be removed entirely. It's `endian.hpp` that includes `endian.ipp`, and not the other way around.

File: cppwamp/include/cppwamp/internal/udsconnector.ipp [hunk 2]
```diff
@@ -9,7 +9,6 @@
 #include "asioconnector.hpp"
 #include "client.hpp"
 #include "config.hpp"
-#include "udsopener.hpp"
```
Comment (ecorm) [inline cppwamp/include/cppwamp/internal/udsconnector.ipp]: Why did you remove? `internal::UdsOpener` is used everywhere in this source file.

Comment (ecorm) [inline cppwamp/include/cppwamp/internal/udsconnector.ipp]: Oh, I see you added `#include "./internal/udsopener.hpp"` in `udsconnector.hpp`. Please see comment for `udsconnector.hpp`.

File: cppwamp/include/cppwamp/internal/endian.ipp (hunk 1 above)
Comment (taion) [inline cppwamp/include/cppwamp/internal/endian.ipp]: Oops, my bad. Wasn't thinking.

File: cppwamp/include/cppwamp/udsconnector.hpp [hunk 3]
```diff
@@ -20,6 +20,7 @@
 #include "connector.hpp"
 #include "error.hpp"
 #include "rawsockdefs.hpp"
+#include "./internal/udsopener.hpp"
```
Comment (ecorm) [inline cppwamp/include/cppwamp/udsconnector.hpp]: This breaks the Pimpl idiom intended by
```
std::unique_ptr<Impl> impl_;
```

You didn't do the same in `tcpconnector.hpp`. Was there a clang error that made you do this? If so, what was the error?

**Comment by ecorm** [comment]
I checked everything except for the tuple stuff. I'm a bit too tired to parse that recursive madness, lol. I'll try going over the tuple stuff tomorrow evening.

Most of it looks fine, and I left comments where I had concerns / questions. Please respond when you get the chance. I see that you already addressed one of them.

We might need more test cases involving `vector<bool>`, but that can be part of a separate issue.

File: cppwamp/include/cppwamp/udsconnector.hpp (hunk 3 above)
Comment (taion) [inline cppwamp/include/cppwamp/udsconnector.hpp]: You're right. I didn't notice the parallel with `tcpconnector.hpp`. It's due to some inconsistency in how clang handles nested classes for forward declarations. I'll have a better fix shortly.

**Comment by taion** [comment]
Accidentally checked in the wrong code for the last commit a couple of times. This last version keeps the correct forwarding semantics on the constructor arguments, and correctly adds the test case to ensure that converting a nested tuple into a nested array still works.

File: cppwamp/include/cppwamp/internal/varianttuple.ipp [hunk 4]
```diff
@@ -14,37 +14,42 @@ namespace wamp
 namespace internal
 {
 
-template<typename TResult, std::size_t N, typename... Ts>
-using TupleResult = typename std::enable_if<N != sizeof...(Ts), TResult>::type;
+//------------------------------------------------------------------------------
```
Comment (ecorm) [inline cppwamp/include/cppwamp/internal/varianttuple.ipp]: This is actually cleaner than what I had before. Kudos!

File: cppwamp/include/cppwamp/internal/varianttuple.ipp [hunk 5]
```diff
@@ -14,37 +14,42 @@ namespace wamp
...
+// This should be possible with return type deduction, but clang doesn't like
+// that for some reason: http://stackoverflow.com/q/24454664.
+template <std::size_t N, typename... Ts>
+using EnableIfTupleElement =
+    typename std::enable_if<N != sizeof...(Ts), int>::type;
```
Comment (ecorm) [inline cppwamp/include/cppwamp/internal/varianttuple.ipp]: This is actually cleaner than what I had before. Kudos!

**Comment by taion** [comment]
Thanks!

**Comment by ecorm** [comment]
Good job, and thanks again for the contribution! :thumbsup:

Do you mind telling me which operating systems you tested this on, so that I can add a few more supported platforms to the project's README? You've already mentioned the compiler versions above.

**Comment by taion** [comment]
OS X 10.10.3 (with clang)
Debian 8.0 "jessie" (with gcc)

I haven't actually built anything using the library yet. I only ran the tests, so I'm not positive everything works when actually using the library, but I plan to get to that in the next week or so.

**Comment by ecorm** [comment]
Thanks. The README says: "This library has been tested with: ...", so its not making any bold claims, hehe.
